postgres-appliance/bootstrap/clone_with_wale.py

```diff
@@ -61,36 +62,114 @@ def fix_output(output):
-    if match is None:
-        raise Exception("wal-e could not found any backups prior to the point in time {0}".format(recovery_target_time))
-    return match['name']
+    if match is not None:
+        return match['name']
+
+
+def list_backups(env):
+    backup_list_cmd = build_wale_command('backup-list')
+    output = subprocess.check_output(backup_list_cmd, env=env)
+    reader = csv.DictReader(fix_output(output), dialect='excel-tab')
+    return list(reader)
+
+
+def get_clone_envdir():
+    from spilo_commons import get_patroni_config
+
+    config = get_patroni_config()
+    restore_command = shlex.split(config['bootstrap']['clone_with_wale']['recovery_conf']['restore_command'])
+    if len(restore_command) > 4 and restore_command[0] == 'envdir':
+        return restore_command[1]
+    raise Exception('Failed to find clone envdir')
+
+
+def get_possible_versions():
+    from spilo_commons import LIB_DIR, get_binary_version, get_bin_dir, get_patroni_config
+
+    config = get_patroni_config()
+
+    max_version = float(get_binary_version(config.get('postgresql', {}).get('bin_dir')))
+
+    versions = {}
+
+    for d in os.listdir(LIB_DIR):
+        try:
+            ver = get_binary_version(get_bin_dir(d))
+            fver = float(ver)
+            if fver <= max_version:
+                versions[fver] = ver
+        except Exception:
+            pass
+
+    # return possible versions in reversed order, i.e. 12, 11, 10, 9.6, and so on
+    return [ver for _, ver in sorted(versions.items(), reverse=True)]
+
+
+def get_wale_environments(env):
+    use_walg = env.get('USE_WALG_RESTORE') == 'true'
+    prefix = 'WALG_' if use_walg else 'WALE_'
+    names = [name for name in env.keys() if name.endswith('_PREFIX') and name.startswith(prefix) and len(name) > 12]
+    if len(names) != 1:
+        raise Exception('Found find {0} {1}*_PREFIX environment variables, expected 1'
+                        .format(len(names), prefix))
+
+    name = names[0]
+    value = env[name].rstrip('/')
+
+    if '/spilo/' in value and value.endswith('/wal'):  # path crafted in the configure_spilo.py?
+        # Try all versions descending if we don't know the version of the source cluster
+        for version in get_possible_versions():
+            yield name, '{0}/{1}/'.format(value, version)
+
+    # Last, try the original value
+    yield name, env[name]
+
+
+def find_backup(recovery_target_time, env):
+    old_value = None
+    for name, value in get_wale_environments(env):
+        if not old_value:
+            old_value = env[name]
+        env[name] = value
+        backup_list = list_backups(env)
+        if backup_list:
+            if recovery_target_time:
+                backup = choose_backup(backup_list, recovery_target_time)
+                if backup:
+                    return backup, (name if value != old_value else None)
+            else:  # We assume that the LATEST backup will be for the biggest postgres version!
```

Comment:
If we don't have the major version of the source cluster specified explicitly we try all postgres versions starting from the biggest. I.e., `get_wale_environments()` function yields tuples:
* ('WALE_S3_PREFIX', 's3://$bucket/spilo/cluster-name/$uid/wal/12')
* ('WALE_S3_PREFIX', 's3://$bucket/spilo/cluster-name/$uid/wal/11')
* ('WALE_S3_PREFIX', 's3://$bucket/spilo/cluster-name/$uid/wal/10')
and so on.

For every prefix we call `wal-e backup-list` and trying to find the backup suitable for the given `recovery_target_time`.
If the `recovery_target_time` is not specified we just pick the `LATEST` backup.

But! It might be, that under the `s3://$bucket/spilo/cluster-name/$uid/wal/` path there are backups for 12 and lets say 10. The correct way of selecting the latest backup between two (or more) different versions would be listing backups for all versions and choosing between them. This is too much work with too few benefits. Therefore I made an assumption if the backup for version 12 is there to not continue with other versions, because most likely the backup for 10 would be older.